// ts-simple-ast/dist/structurePrinters/function/FunctionDeclarationStructurePrinter.js
"use strict";
var __extends = (this && this.__extends)/* istanbul ignore next */ || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values)/* istanbul ignore next */ || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../utils");
var StructurePrinter_1 = require("../StructurePrinter");
var ParameterDeclarationStructurePrinter_1 = require("./ParameterDeclarationStructurePrinter");
var base_1 = require("../base");
var doc_1 = require("../doc");
var formatting_1 = require("../formatting");
var types_1 = require("../types");
var statement_1 = require("../statement");
var FunctionDeclarationStructurePrinter = /** @class */ (function (_super) {
    __extends(FunctionDeclarationStructurePrinter, _super);
    function FunctionDeclarationStructurePrinter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.jsDocWriter = new doc_1.JSDocStructurePrinter(_this.writer);
        _this.modifierWriter = new base_1.ModifierableNodeStructurePrinter(_this.writer);
        _this.typeParameterWriter = new types_1.TypeParameterDeclarationStructurePrinter(_this.writer);
        _this.parametersWriter = new formatting_1.CommaSeparatedStructuresPrinter(_this.writer, new ParameterDeclarationStructurePrinter_1.ParameterDeclarationStructurePrinter(_this.writer));
        return _this;
    }
    FunctionDeclarationStructurePrinter.prototype.printTexts = function (structures) {
        if (structures == null)
            return;
        for (var i = 0; i < structures.length; i++) {
            var currentStructure = structures[i];
            if (i > 0) {
                var previousStructure = structures[i - 1];
                if (previousStructure.hasDeclareKeyword && currentStructure.hasDeclareKeyword)
                    this.writer.newLine();
                else
                    this.writer.blankLine();
            }
            this.printText(currentStructure);
        }
    };
    FunctionDeclarationStructurePrinter.prototype.printText = function (structure) {
        var _this = this;
        this.printOverloads(structure.name, getOverloadStructures());
        this.printBase(structure.name, structure);
        if (structure.hasDeclareKeyword)
            this.writer.write(";");
        else
            this.writer.space().inlineBlock(function () {
                new statement_1.BodyTextStructurePrinter(_this.writer, { isAmbient: false }).printText(structure);
            });
        function getOverloadStructures() {
            // all the overloads need to have similar properties as the implementation
            var overloads = utils_1.ObjectUtils.clone(structure.overloads);
            if (overloads == null || overloads.length === 0)
                return;
            try {
                for (var overloads_1 = __values(overloads), overloads_1_1 = overloads_1.next(); !overloads_1_1.done; overloads_1_1 = overloads_1.next()) {
                    var overload = overloads_1_1.value;
                    utils_1.setValueIfUndefined(overload, "hasDeclareKeyword", structure.hasDeclareKeyword);
                    utils_1.setValueIfUndefined(overload, "isExported", structure.isExported);
                    utils_1.setValueIfUndefined(overload, "isDefaultExport", structure.isDefaultExport);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (overloads_1_1 && !overloads_1_1.done && (_a = overloads_1.return)) _a.call(overloads_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return overloads;
            var e_1, _a;
        }
    };
    FunctionDeclarationStructurePrinter.prototype.printOverloads = function (name, structures) {
        if (structures == null || structures.length === 0)
            return;
        try {
            for (var structures_1 = __values(structures), structures_1_1 = structures_1.next(); !structures_1_1.done; structures_1_1 = structures_1.next()) {
                var structure = structures_1_1.value;
                this.printOverload(name, structure);
                this.writer.newLine();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (structures_1_1 && !structures_1_1.done && (_a = structures_1.return)) _a.call(structures_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        var e_2, _a;
    };
    FunctionDeclarationStructurePrinter.prototype.printOverload = function (name, structure) {
        this.printBase(name, structure);
        this.writer.write(";");
    };
    FunctionDeclarationStructurePrinter.prototype.printBase = function (name, structure) {
        this.jsDocWriter.printDocs(structure.docs);
        this.modifierWriter.printText(structure);
        this.writer.write("function");
        this.writer.conditionalWrite(structure.isGenerator, "*");
        this.writer.write(" " + name);
        this.typeParameterWriter.printTexts(structure.typeParameters);
        this.writer.write("(");
        if (structure.parameters != null)
            this.parametersWriter.printText(structure.parameters);
        this.writer.write(")");
        if (!utils_1.StringUtils.isNullOrWhitespace(structure.returnType))
            this.writer.write(": " + structure.returnType);
    };
    return FunctionDeclarationStructurePrinter;
}(StructurePrinter_1.StructurePrinter));
exports.FunctionDeclarationStructurePrinter = FunctionDeclarationStructurePrinter;
