// ts-simple-ast/dist/compiler/literal/template/TaggedTemplateExpression.js
"use strict";
var __extends = (this && this.__extends)/* istanbul ignore next */ || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var manipulation_1 = require("../../../manipulation");
var expression_1 = require("../../expression");
exports.TaggedTemplateExpressionBase = expression_1.MemberExpression;
var TaggedTemplateExpression = /** @class */ (function (_super) {
    __extends(TaggedTemplateExpression, _super);
    function TaggedTemplateExpression() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets the tag.
     */
    TaggedTemplateExpression.prototype.getTag = function () {
        return this.getNodeFromCompilerNode(this.compilerNode.tag);
    };
    /**
     * Gets the template literal.
     */
    TaggedTemplateExpression.prototype.getTemplate = function () {
        return this.getNodeFromCompilerNode(this.compilerNode.template);
    };
    /**
     * Removes the tag from the tagged template.
     * @returns The new template expression.
     */
    TaggedTemplateExpression.prototype.removeTag = function () {
        var parent = this.getParentSyntaxList() || this.getParentOrThrow();
        var index = this.getChildIndex();
        var template = this.getTemplate();
        manipulation_1.insertIntoParentTextRange({
            customMappings: function (newParent) { return [{ currentNode: template, newNode: newParent.getChildren()[index] }]; },
            parent: parent,
            insertPos: this.getStart(),
            newText: this.getTemplate().getText(),
            replacing: {
                textLength: this.getWidth(),
                nodes: [this]
            }
        });
        return parent.getChildAtIndex(index);
    };
    return TaggedTemplateExpression;
}(exports.TaggedTemplateExpressionBase));
exports.TaggedTemplateExpression = TaggedTemplateExpression;
