// ts-simple-ast/dist/compiler/jsx/JsxSpreadAttribute.js
"use strict";
var __extends = (this && this.__extends)/* istanbul ignore next */ || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var manipulation_1 = require("../../manipulation");
var common_1 = require("../common");
var JsxSpreadAttribute = /** @class */ (function (_super) {
    __extends(JsxSpreadAttribute, _super);
    function JsxSpreadAttribute() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets the JSX spread attribute's expression.
     */
    JsxSpreadAttribute.prototype.getExpression = function () {
        return this.getNodeFromCompilerNode(this.compilerNode.expression);
    };
    /**
     * Removes the JSX spread attribute.
     */
    JsxSpreadAttribute.prototype.remove = function () {
        manipulation_1.removeChildren({
            children: [this],
            removePrecedingNewLines: true,
            removePrecedingSpaces: true
        });
    };
    return JsxSpreadAttribute;
}(common_1.Node));
exports.JsxSpreadAttribute = JsxSpreadAttribute;
