// ts-simple-ast/dist/fileSystem/DirectoryCoordinator.js
"use strict";
var __values = (this && this.__values)/* istanbul ignore next */ || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var errors = require("../errors");
var utils_1 = require("../utils");
/**
 * Contains common methods between Project and Directory.
 *
 * I'll definitely need to refactor this in the future... just putting these methods in a common place for now.
 */
var DirectoryCoordinator = /** @class */ (function () {
    function DirectoryCoordinator(compilerFactory, fileSystemWrapper) {
        this.compilerFactory = compilerFactory;
        this.fileSystemWrapper = fileSystemWrapper;
    }
    DirectoryCoordinator.prototype.addExistingDirectoryIfExists = function (dirPath, options) {
        var directory = this.compilerFactory.getDirectoryFromPath(dirPath);
        if (directory == null)
            return undefined;
        if (options.recursive) {
            try {
                for (var _a = __values(utils_1.FileUtils.getDescendantDirectories(this.fileSystemWrapper, dirPath)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var descendantDirPath = _b.value;
                    this.compilerFactory.createOrAddDirectoryIfNotExists(descendantDirPath);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return directory;
        var e_1, _c;
    };
    DirectoryCoordinator.prototype.addExistingDirectory = function (dirPath, options) {
        var directory = this.addExistingDirectoryIfExists(dirPath, options);
        if (directory == null)
            throw new errors.DirectoryNotFoundError(dirPath);
        return directory;
    };
    DirectoryCoordinator.prototype.createDirectory = function (dirPath) {
        return this.compilerFactory.createDirectory(dirPath);
    };
    return DirectoryCoordinator;
}());
exports.DirectoryCoordinator = DirectoryCoordinator;
