// ts-simple-ast/dist/utils/references/LazyReferenceCoordinator.js
"use strict";
var __values = (this && this.__values)/* istanbul ignore next */ || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var collections_1 = require("../collections");
/**
 * Updates all the source file's reference containers.
 */
var LazyReferenceCoordinator = /** @class */ (function () {
    function LazyReferenceCoordinator(factory) {
        var _this = this;
        this.dirtySourceFiles = collections_1.createHashSet();
        factory.onSourceFileAdded(function (sourceFile) {
            _this.dirtySourceFiles.add(sourceFile);
            sourceFile.onModified(function () {
                if (!sourceFile.wasForgotten())
                    _this.dirtySourceFiles.add(sourceFile);
            });
        });
        factory.onSourceFileRemoved(function (sourceFile) {
            sourceFile._referenceContainer.clear();
            _this.dirtySourceFiles.delete(sourceFile);
        });
    }
    LazyReferenceCoordinator.prototype.refreshDirtySourceFiles = function () {
        try {
            for (var _a = __values(this.dirtySourceFiles.values()), _b = _a.next(); !_b.done; _b = _a.next()) {
                var sourceFile = _b.value;
                sourceFile._referenceContainer.refresh();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.clearDirtySourceFiles();
        var e_1, _c;
    };
    LazyReferenceCoordinator.prototype.refreshSourceFileIfDirty = function (sourceFile) {
        if (!this.dirtySourceFiles.has(sourceFile))
            return;
        sourceFile._referenceContainer.refresh();
        this.clearDityForSourceFile(sourceFile);
    };
    LazyReferenceCoordinator.prototype.clearDirtySourceFiles = function () {
        this.dirtySourceFiles.clear();
    };
    LazyReferenceCoordinator.prototype.clearDityForSourceFile = function (sourceFile) {
        this.dirtySourceFiles.delete(sourceFile);
    };
    return LazyReferenceCoordinator;
}());
exports.LazyReferenceCoordinator = LazyReferenceCoordinator;
